File: source/producers/fakeTwitterProducer.py
import sys
from kafka import KafkaProducer
from producer import Producer
import random
import time


class FakeTwitterProducer(Producer):
    rate = 1.0 # tweets/sec
    allowance = rate
    last_check = time.clock()

    def __init__(self, rate, logging=False):
        Producer.__init__(self, logging)
        self.fake_tweet = open('fake_tweet.json', 'r').read()
        self.rate = rate

    def generate_tweet(self):
        return self.fake_tweet

    def run_producer(self):
        while True:
            current = time.clock()
            time_passed = current - self.last_check
            if time_passed >= self.rate:
                self.send_to_kafka(self.generate_tweet())
                self.last_check = current


if __name__ == "__main__":
    logging = False
    if len(sys.argv) > 3:
        print "Usage : python fake_twitter [--log]"
        exit(1)
    if '--log' in sys.argv:
        logging = True
        sys.argv.remove('--log')
    rate = sys.argv[1]

    FakeTwitterProducer(
        1/float(rate),
        logging
    ).run_producer()

File: source/flink-hbase/src/main/java/org/apache/flink/quickstart/FakeTwitterProducer.java
package org.apache.flink.quickstart;

import com.sun.org.apache.xpath.internal.functions.WrongNumberArgsException;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Properties;

/**
 * Created by willol on 08/12/16.
 */
public class FakeTwitterProducer extends KafkaProducer<String, String> {
	private int rate;
	private double lastCheck;
	private String fakeTweet;

	public FakeTwitterProducer(Properties props, int rate) throws IOException {
		super(props);
		this.rate = rate;
		this.lastCheck = System.currentTimeMillis() / 1000.0;

		/* read the fake tweet */
		File file = new File("src/main/java/org/apache/flink/quickstart/fake_tweet.json");
		FileInputStream fis = new FileInputStream(file);
		byte[] data = new byte[(int) file.length()];
		fis.read(data);
		fis.close();
		this.fakeTweet = new String(data, "UTF-8");
	}

	public String generateTweet() {
		return this.fakeTweet;
	}

	public void runProducer() {
		while (true) {
			double current = System.currentTimeMillis() / 1000.0;
			double timePassed = current - this.lastCheck;
			if (timePassed >= this.rate) {
				this.sendToKafka(this.generateTweet());
				this.lastCheck = current;
			}
		}
	}

	public void sendToKafka(String data) {
		this.send(new ProducerRecord<String, String>(
				"twitter",
				data
		));
	}

	public static void main(String[] argv) throws IOException, WrongNumberArgsException {
		if (argv.length < 2) {
			throw new WrongNumberArgsException("Arg 1 : rate\nArg 2 : kafka_address:9092");
		}
		int rate = Integer.parseInt(argv[0]);

		/* Kafka properties */
		Properties props = new Properties();
		props.put("bootstrap.servers", argv[1]);
		props.put("acks", "all");
		props.put("retries", 0);
		props.put("batch.size", 16384);
		props.put("linger.ms", 1);
		props.put("buffer.memory", 33554432);
		props.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");
		props.put("value.serializer", "org.apache.kafka.common.serialization.StringSerializer");

		FakeTwitterProducer provider = new FakeTwitterProducer(props, rate);
		provider.runProducer();
	}
}
